### ISSUE TITLE
Bug: Constraint disregarded in very peculiar cases.

### ISSUE THREAD

**Issue by stian-svedenborg** [MiniZinc/libminizinc#42]
Hi, I have encountered an interesting bug. In the following model, the constraint is completely disregarded, unless I use the function elsewhere.

```
set of int: idx_set = 1..2;

array[idx_set] of int: t1 = [0, 2];
array[idx_set] of int: t2 = [0, -2];

var 0..1: negative;

function var int:
select_negative(var int: idx) = t2[idx];

function var int:
select_positive(var int: idx) = t1[idx];


function var int:
select(var int: idx) =
	if negative == 0 then
		select_positive(idx)
	else
		select_negative(idx)
	endif
	;

var idx_set: idx;
var -2..2: result;

constraint ( ( select(idx) != 0) -> (result == select(idx)) );

% Uncommenting any of these blocks makes the issue disappear.
% Block 1
% var int: result2;
% constraint ( result2 == select(idx));

% Block 2
%constraint ( ( select_positive(idx) != 0) -> (result == select(idx)) );

% Block 3
% Replacing the constraint above with this also works.
%constraint (
%    let { var int: x = select(idx) } in
%    ( x != 0) -> (result == x)
%);


solve satisfy;
```

When solved with fzn-gecode -a I expect there to be exactly two solutions for idx == 2, i.e. one for negative == 1 and one for negative == 0. Uncommenting any of the listed blocks produces the expected output.

**Comment by guidotack** [comment]
Sorry for the late reply. This is indeed a bug, the problem is the combination of if-then-else, functions and array access expressions. I'm working on a fix.

**Comment by guidotack** [comment]
Fixed by 08a622cb0619eba3c4fafac5e0e20cf4a8522b50.
